gcloud-java-storage/src/test/java/com/google/gcloud/storage/ITStorageTest.java

```diff
@@ -0,0 +1,470 @@
+    BlobInfo remoteBlob = storage.create(blob, BLOB_BYTE_CONTENT);
+    assertNotNull(remoteBlob);
+    assertEquals(blob.bucket(), remoteBlob.bucket());
+    assertEquals(blob.name(), remoteBlob.name());
+    byte[] readBytes = storage.readAllBytes(bucket, blobName);
+    assertArrayEquals(BLOB_BYTE_CONTENT, readBytes);
+    assertTrue(storage.delete(bucket, blobName));
+  }
+
+  @Test
+  public void testCreateEmptyBlob() {
+    String blobName = "test-create-empty-blob";
+    BlobInfo blob = BlobInfo.of(bucket, blobName);
+    BlobInfo remoteBlob = storage.create(blob);
+    assertNotNull(remoteBlob);
+    assertEquals(blob.bucket(), remoteBlob.bucket());
+    assertEquals(blob.name(), remoteBlob.name());
+    byte[] readBytes = storage.readAllBytes(bucket, blobName);
+    assertArrayEquals(new byte[0], readBytes);
+    assertTrue(storage.delete(bucket, blobName));
+  }
+
+  @Test
+  public void testCreateBlobStream() throws UnsupportedEncodingException {
+    String blobName = "test-create-blob-stream";
+    BlobInfo blob = BlobInfo.builder(bucket, blobName).contentType(CONTENT_TYPE).build();
+    ByteArrayInputStream stream = new ByteArrayInputStream(BLOB_STRING_CONTENT.getBytes(UTF_8));
+    BlobInfo remoteBlob = storage.create(blob, stream);
+    assertNotNull(remoteBlob);
+    assertEquals(blob.bucket(), remoteBlob.bucket());
+    assertEquals(blob.name(), remoteBlob.name());
+    assertEquals(blob.contentType(), remoteBlob.contentType());
+    byte[] readBytes = storage.readAllBytes(bucket, blobName);
+    assertEquals(BLOB_STRING_CONTENT, new String(readBytes, UTF_8));
+    assertTrue(storage.delete(bucket, blobName));
+  }
+
+  @Test
+  public void testCreateBlobFail() {
+    String blobName = "test-create-blob-fail";
+    BlobInfo blob = BlobInfo.of(bucket, blobName);
+    assertNotNull(storage.create(blob));
+    try {
+      storage.create(blob.toBuilder().generation(42L).build(), BLOB_BYTE_CONTENT,
+          Storage.BlobTargetOption.generationMatch());
+      fail("StorageException was expected");
+    } catch (StorageException ex) {
+      // expected
+    }
+    assertTrue(storage.delete(bucket, blobName));
+  }
+
+  @Test
+  public void testUpdateBlob() {
+    String blobName = "test-update-blob";
+    BlobInfo blob = BlobInfo.of(bucket, blobName);
+    assertNotNull(storage.create(blob));
+    BlobInfo updatedBlob = storage.update(blob.toBuilder().contentType(CONTENT_TYPE).build());
+    assertNotNull(updatedBlob);
+    assertEquals(blob.bucket(), updatedBlob.bucket());
+    assertEquals(blob.name(), updatedBlob.name());
+    assertEquals(CONTENT_TYPE, updatedBlob.contentType());
+    assertTrue(storage.delete(bucket, blobName));
+  }
+
+  @Test
+  public void testUpdateBlobFail() {
+    String blobName = "test-update-blob-fail";
+    BlobInfo blob = BlobInfo.of(bucket, blobName);
+    assertNotNull(storage.create(blob));
+    try {
+      storage.update(blob.toBuilder().contentType(CONTENT_TYPE).generation(42L).build(),
```

Comment:
Feel free to leave as is. I don't think GCS promise anything about the generation number (and theoretically could a first value could being with 42).

Fine as is, though you may also check that 42 didn't happen to be the right one (and if so use any other valid value).